ee/modules/030-cloud-provider-openstack/hooks/discover_test.go

```diff
@@ -0,0 +1,259 @@
+apiVersion: v1
+kind: Secret
+metadata:
+  name: d8-cloud-provider-discovery-data
+  namespace: kube-system
+data:
+  "discovery-data.json": %s
+`, base64.StdEncoding.EncodeToString([]byte(discoveryData)))
+
+	a := HookExecutionConfigInit(initValues, `{}`)
+	Context("Cluster has empty state", func() {
+		BeforeEach(func() {
+			a.BindingContexts.Set(a.KubeStateSet(``))
+			a.RunHook()
+		})
+
+		It("Hook should not fail with errors", func() {
+			Expect(a).To(ExecuteSuccessfully())
+			Expect(a.GoHookError).Should(BeNil())
+		})
+	})
+
+	b := HookExecutionConfigInit(initValues, `{}`)
+	Context("Provider data is successfully discovered", func() {
+		BeforeEach(func() {
+			b.BindingContexts.Set(b.KubeStateSet(state))
+			b.RunHook()
+		})
+
+		It("Zones values should be gathered from discovered data", func() {
+			Expect(b).To(ExecuteSuccessfully())
+
+			Expect(b.ValuesGet("cloudProviderOpenstack.internal.discoveryData.zones").String()).To(MatchJSON(`["nova", "zz-zzz-1z", "xx-xxx-1x", "cc-ccc-1c"]`))
+		})
+
+		It("Should discover all volumeTypes and no default", func() {
+			Expect(b).To(ExecuteSuccessfully())
+			Expect(b.ValuesGet("cloudProviderOpenstack.internal.storageClasses").String()).To(MatchJSON(`
+[
+  {
+	"name": "ceph-ssd",
+	"type": "ceph-ssd"
+  },
+  {
+	"name": "default",
+	"type": "__DEFAULT__"
+  },
+  {
+	"name": "other-bar",
+	"type": "other-bar"
+  },
+  {
+	"name": "some-foo",
+	"type": "some-foo"
+  },
+  {
+	"name": "ssd-r1",
+	"type": "SSD R1"
+  },
+  {
+	"name": "xx--foo",
+	"type": "-Xx__$()? -foo-"
+  },
+  {
+	"name": "yy-fast-ssd-foo",
+	"type": "  YY fast SSD-foo."
+  }
+]
+`))
+			Expect(b.ValuesGet("cloudProviderOpenstack.internal.defaultStorageClass").Exists()).To(BeFalse())
+		})
+	})
+
+	initValues = `
+cloudProviderOpenstack:
+  internal:
+    connection:
+      authURL: https://test.tests.com:5000/v3/
+      username: jamie
+      password: nein
+      domainName: default
+      tenantName: default
+      tenantID: "123"
+      region: HetznerFinland
+  storageClass:
+    exclude:
+    - .*-foo
+    - bar
+    default: other-bar
+`
+
+	c := HookExecutionConfigInit(initValues, `{}`)
+	Context("Provider data is successfully discovered", func() {
+		BeforeEach(func() {
+			c.BindingContexts.Set(c.KubeStateSet(state))
+			c.RunHook()
+		})
+
+		It("All values should be gathered from discovered data", func() {
+			Expect(c).To(ExecuteSuccessfully())
+		})
+
+		It("Should discover volumeTypes without excluded and default set", func() {
+			Expect(c).To(ExecuteSuccessfully())
+			Expect(c.ValuesGet("cloudProviderOpenstack.internal.storageClasses").String()).To(MatchJSON(`
+[
+  {
+	"name": "ceph-ssd",
+	"type": "ceph-ssd"
+  },
+  {
+	"name": "default",
+	"type": "__DEFAULT__"
+  },
+  {
+	"name": "other-bar",
+	"type": "other-bar"
+  },
+  {
+	"name": "ssd-r1",
+	"type": "SSD R1"
+  }
+]
+`))
+			Expect(c.ValuesGet("cloudProviderOpenstack.internal.defaultStorageClass").String()).To(Equal(`other-bar`))
+		})
+	})
```

Comment:
Need to add test with no discovery data secret and:
   no storage classes - hook should not fail
   have storage classes - storage classes should in  values